File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/459-repeated-substring-pattern.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func repeatedSubstringPattern(s string) bool {
	// TODO: implement
	return false
}

func runTests() {
	assert(repeatedSubstringPattern("abab") == true, "test 1")
	assert(repeatedSubstringPattern("aba") == false, "test 2")
	assert(repeatedSubstringPattern("abcabcabcabc") == true, "test 3")
	assert(repeatedSubstringPattern("a") == false, "test 4")
	assert(repeatedSubstringPattern("aa") == true, "test 5")
	assert(repeatedSubstringPattern("abaaba") == true, "test 6")
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/459-repeated-substring-pattern.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function repeatedSubstringPattern(s: string): boolean {
    // TODO: implement
    return false;
}

function _runTests(): void {
    assert(repeatedSubstringPattern('abab') === true);
    assert(repeatedSubstringPattern('aba') === false);
    assert(repeatedSubstringPattern('abcabcabcabc') === true);
    assert(repeatedSubstringPattern('a') === false);
    assert(repeatedSubstringPattern('aa') === true);
    assert(repeatedSubstringPattern('abaaba') === true);
    // perf
    const t0 = performance.now();
    repeatedSubstringPattern('ab'.repeat(5000));
    console.log(`perf repeatedSubstringPattern n=10000: ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
